### PR TITLE
369 Enhance segmentation loss testing

### LINKED ISSUE #369: Incorrect implementation of generalized Dice loss
In the line 195 of dice.py the element-wise division cancels out the weights (if `smooth` is negligible), which destroys the most important property of generalized Dice loss. 

https://github.com/Project-MONAI/MONAI/blob/ad06dff7f85711048690b2e85c99d51001612708/monai/losses/dice.py#L195

### PR DESCRIPTION
**Pull request by wyli**
### Description
A few enhancements for the segmentation losses:
- fixes #369 
- inheriting `_Loss`: forward argument names changed `pred`->`input`, `ground`->`target`
- inheriting `_Loss`: adds `reduction` argument
- adds integration tests for all segmentation losses

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -78,36 +83,39 @@ def forward(self, pred, ground, smooth=1e-5):
...
 
         denominator = ground_o + pred_o
 
         if self.jaccard:
             denominator -= intersection
 
-        f = (2.0 * intersection + smooth) / (denominator + smooth)
-        return 1.0 - f.mean()  # final reduce_mean across batches and channels
+        f = 1.0 - (2.0 * intersection + smooth) / (denominator + smooth)
+        if self.reduction == "sum":
```
Comment (Nic-Ma) [inline monai/losses/dice.py:114]: Suggest to simply this `reduction` logic according to PyTorch [source code](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2613) as all loss functions need it:
```py
if reduction != 'none':
    ret = torch.mean(ret) if reduction == 'mean' else torch.sum(ret)
```
Thanks.

File: monai/losses/dice.py [hunk 2]
```diff
@@ -192,8 +210,12 @@ def forward(self, pred, ground, smooth=1e-5):
             b[infs] = 0.0
             b[infs] = torch.max(b)
 
-        f = (2.0 * intersection * w + smooth) / (denominator * w + smooth)
-        return 1.0 - f.mean()  # final reduce_mean across batches and channels
+        f = 1.0 - (2.0 * (intersection * w).sum(1) + smooth) / ((denominator * w).sum(1) + smooth)
```
Comment (Nic-Ma) [inline monai/losses/dice.py:213]: same as above comment.

File: monai/losses/focal_loss.py [hunk 3]
```diff
@@ -91,8 +94,7 @@ def forward(self, input, target):
 
         if self.reduction == "sum":
             return loss.sum()
-        elif self.reduction == "none":
+        if self.reduction == "none":
```
Comment (Nic-Ma) [inline monai/losses/focal_loss.py:97]: same as above comment.

File: monai/losses/tversky.py [hunk 4]
```diff
@@ -101,6 +113,10 @@ def forward(self, pred, ground, smooth=1e-5):
         numerator = tp + smooth
         denominator = tp + fp + fn + smooth
 
-        score = numerator / denominator
+        score = 1.0 - numerator / denominator
 
-        return 1.0 - score.mean()
+        if self.reduction == "sum":
```
Comment (Nic-Ma) [inline monai/losses/tversky.py:118]: same as above comment.

File: monai/losses/dice.py (hunk 1 above)
Comment (wyli) [reply monai/losses/dice.py:114]: thanks, there are different styles in the link you provided, now following the robust one https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2756

Comment (Nic-Ma) [reply monai/losses/dice.py:114]: Looks good.
Thanks.
